File: backend/app/schemas/account.py
"""
Apple ID 账号相关的 Pydantic Schemas
"""
from datetime import datetime
from typing import Optional
from pydantic import BaseModel, EmailStr, Field


# ========== Apple ID 创建 ==========
class AppleAccountCreate(BaseModel):
    """创建 Apple ID Schema"""
    apple_id: EmailStr = Field(..., description="Apple ID（邮箱）")
    password: str = Field(..., min_length=1, description="密码")
    security_questions: Optional[list[dict]] = Field(None, description="密保问题")
    tags: Optional[list[str]] = Field(None, description="标签")
    category: Optional[str] = Field(None, max_length=50, description="分类")
    note: Optional[str] = Field(None, description="备注")

    # 自动化配置
    auto_unlock: bool = Field(True, description="是否自动解锁")
    auto_disable_2fa: bool = Field(False, description="是否自动关闭2FA")
    auto_change_password: bool = Field(False, description="是否自动修改密码")
    auto_remove_devices: bool = Field(False, description="是否自动删除设备")
    check_interval: int = Field(3600, ge=300, description="检测间隔（秒）")

    class Config:
        json_schema_extra = {
            "example": {
                "apple_id": "example@icloud.com",
                "password": "password123",
                "security_questions": [
                    {
                        "question": "您最喜欢的颜色是什么?",
                        "answer": "蓝色"
                    }
                ],
                "tags": ["工作", "美区"],
                "auto_unlock": True,
                "check_interval": 3600
            }
        }


# ========== Apple ID 更新 ==========
class AppleAccountUpdate(BaseModel):
    """更新 Apple ID Schema"""
    password: Optional[str] = Field(None, description="密码")
    security_questions: Optional[list[dict]] = Field(None, description="密保问题")
    tags: Optional[list[str]] = Field(None, description="标签")
    category: Optional[str] = Field(None, description="分类")
    note: Optional[str] = Field(None, description="备注")
    auto_unlock: Optional[bool] = Field(None, description="是否自动解锁")
    auto_disable_2fa: Optional[bool] = Field(None, description="是否自动关闭2FA")
    auto_change_password: Optional[bool] = Field(None, description="是否自动修改密码")
    auto_remove_devices: Optional[bool] = Field(None, description="是否自动删除设备")
    check_interval: Optional[int] = Field(None, ge=300, description="检测间隔（秒）")


# ========== Apple ID 响应 ==========
class AppleAccountResponse(BaseModel):
    """Apple ID 响应 Schema"""
    id: int
    apple_id: str
    status: str
    lock_status: bool
    two_factor_enabled: bool
    app_store_country: Optional[str]
    app_store_language: Optional[str]
    account_name: Optional[str]
    devices_count: int
    has_lost_mode: bool
    tags: Optional[list[str]]
    category: Optional[str]
    auto_unlock: bool
    auto_disable_2fa: bool
    auto_change_password: bool
    auto_remove_devices: bool
    check_interval: int
    last_checked_at: Optional[datetime]
    next_check_at: Optional[datetime]
    last_unlocked_at: Optional[datetime]
    unlock_count: int
    last_error: Optional[str]
    error_count: int
    note: Optional[str]
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


# ========== Apple ID 详细响应（含密码）==========
class AppleAccountDetailResponse(AppleAccountResponse):
    """Apple ID 详细响应（含密码）"""
    current_password: Optional[str] = None
    security_questions: Optional[list[dict]] = None


# ========== 密码历史响应 ==========
class PasswordHistoryResponse(BaseModel):
    """密码历史响应 Schema"""
    id: int
    password: str
    changed_by: str
    created_at: datetime

    class Config:
        from_attributes = True


# ========== 批量导入 ==========
class AppleAccountImport(BaseModel):
    """批量导入 Apple ID Schema"""
    accounts: list[dict] = Field(..., description="账号列表")

    class Config:
        json_schema_extra = {
            "example": {
                "accounts": [
                    {
                        "apple_id": "example1@icloud.com",
                        "password": "password123",
                        "tags": ["工作", "美区"],
                        "note": "测试账号1"
                    },
                    {
                        "apple_id": "example2@icloud.com",
                        "password": "password456",
                        "tags": ["个人"],
                        "note": "测试账号2"
                    }
                ]
            }
        }


# ========== 批量导入响应 ==========
class AppleAccountImportResponse(BaseModel):
    """批量导入响应 Schema"""
    total: int
    success: int
    failed: int
    errors: list[dict] = []
    message: str = "导入完成"


# ========== 账号列表查询 ==========
class AppleAccountListQuery(BaseModel):
    """账号列表查询 Schema"""
    page: int = Field(1, ge=1, description="页码")
    per_page: int = Field(20, ge=1, le=100, description="每页数量")
    status: Optional[str] = Field(None, description="状态过滤")
    lock_status: Optional[bool] = Field(None, description="锁定状态过滤")
    tag: Optional[str] = Field(None, description="标签过滤")
    category: Optional[str] = Field(None, description="分类过滤")
    search: Optional[str] = Field(None, description="搜索关键词")


# ========== 手动触发检测 ==========
class TriggerCheck(BaseModel):
    """手动触发检测 Schema"""
    force: bool = Field(False, description="是否强制检测（忽略间隔限制）")

File: frontend/src/types/account.ts
/**
 * Apple ID 账号相关类型定义
 *
 * 职责：
 * - 定义 Apple ID 账号信息类型
 * - 提供账号管理相关的请求和响应类型
 * - 定义账号状态、安全问题等相关类型
 */

import type { PaginationParams, PaginationResponse, ImportResult } from './common'

/**
 * Apple ID 账号状态
 */
export type AccountStatus = 'normal' | 'locked' | 'checking' | 'failed' | 'disabled'

/**
 * Apple ID 账号信息
 */
export interface AppleAccount {
  id: number
  user_id: number
  apple_id: string
  password: string  // 加密后的密码
  status: AccountStatus
  lock_status: boolean  // 是否被锁定
  two_factor_enabled: boolean  // 是否启用了2FA
  // 账号详情
  app_store_country: string  // App Store 国家/地区
  app_store_language?: string  // App Store 语言
  account_name?: string  // 账号姓名
  birthday?: string  // 生日
  phone_number?: string  // 手机号（部分）
  // 设备信息
  devices_count: number  // 设备数量
  has_lost_mode: boolean  // 是否有丢失模式
  trusted_devices?: string[]  // 受信任设备列表
  // 分类和标签
  tags: string[]  // 标签（如：工作、美区、付费等）
  category?: string  // 分类
  note?: string  // 备注
  // 自动化设置
  auto_unlock: boolean  // 是否自动解锁
  auto_disable_2fa: boolean  // 是否自动关闭2FA
  check_interval: number  // 检测间隔（秒）
  priority: number  // 优先级（0-10）
  // 统计信息
  unlock_count: number  // 解锁次数
  check_count: number  // 检测次数
  last_checked_at: string | null  // 最后检测时间
  next_check_at: string | null  // 下次检测时间
  last_unlocked_at: string | null  // 最后解锁时间
  last_error?: string  // 最后错误信息
  // 时间戳
  created_at: string
  updated_at: string
}

/**
 * 安全问题
 */
export interface SecurityQuestion {
  question: string
  answer: string
}

/**
 * 账号列表查询参数
 */
export interface AccountListParams extends PaginationParams {
  status?: AccountStatus
  lock_status?: boolean
  two_factor_enabled?: boolean
  tag?: string
  category?: string
  search?: string  // 搜索 apple_id
  app_store_country?: string
  sort_by?: 'created_at' | 'updated_at' | 'last_checked_at' | 'unlock_count'
  sort_order?: 'asc' | 'desc'
}

/**
 * 账号列表响应
 */
export interface AccountListResponse extends PaginationResponse<AppleAccount> {
  stats?: {
    total: number
    normal: number
    locked: number
    checking: number
    failed: number
  }
}

/**
 * 创建账号请求
 */
export interface CreateAccountRequest {
  apple_id: string
  password: string
  security_questions?: SecurityQuestion[]
  tags?: string[]
  category?: string
  note?: string
  auto_unlock?: boolean
  auto_disable_2fa?: boolean
  check_interval?: number
  priority?: number
}

/**
 * 更新账号请求
 */
export interface UpdateAccountRequest {
  password?: string
  security_questions?: SecurityQuestion[]
  tags?: string[]
  category?: string
  note?: string
  auto_unlock?: boolean
  auto_disable_2fa?: boolean
  check_interval?: number
  priority?: number
  status?: AccountStatus
}

/**
 * 批量导入请求
 */
export interface BatchImportRequest {
  file?: File | Blob
  format: 'csv' | 'json'
  data?: string  // CSV 或 JSON 字符串
  auto_unlock?: boolean
  check_interval?: number
  tags?: string[]
}

/**
 * 批量导入响应
 */
export interface BatchImportResponse extends ImportResult {
  imported_ids?: number[]
}

/**
 * 密码历史记录
 */
export interface PasswordHistory {
  id: number
  account_id: number
  password: string
  changed_by: 'manual' | 'auto' | 'system'
  changed_by_user?: string
  note?: string
  created_at: string
}

/**
 * 账号检测结果
 */
export interface AccountCheckResult {
  account_id: number
  apple_id: string
  status: AccountStatus
  lock_status: boolean
  two_factor_enabled: boolean
  devices_count: number
  has_lost_mode: boolean
  app_store_country: string
  checked_at: string
  error?: string
}

/**
 * 账号详情（包含完整信息）
 */
export interface AccountDetail extends AppleAccount {
  security_questions?: SecurityQuestion[]
  password_history?: PasswordHistory[]
  recent_tasks?: any[]  // 最近的任务记录
  device_list?: Device[]  // 设备列表
}

/**
 * 设备信息
 */
export interface Device {
  id: string
  name: string
  model: string
  os_version: string
  is_trusted: boolean
  last_seen_at: string
}

/**
 * 账号标签统计
 */
export interface AccountTagStats {
  tag: string
  count: number
  normal_count: number
  locked_count: number
}

/**
 * 账号分类统计
 */
export interface AccountCategoryStats {
  category: string
  count: number
  unlock_count: number
}

/**
 * 账号国家/地区统计
 */
export interface AccountCountryStats {
  country: string
  country_name: string
  count: number
  locked_count: number
}

/**
 * 批量操作账号请求
 */
export interface BatchAccountOperation {
  account_ids: number[]
  action: 'unlock' | 'check' | 'disable_2fa' | 'delete' | 'update_tags'
  params?: {
    tags?: string[]
    auto_unlock?: boolean
    check_interval?: number
  }
}

/**
 * 批量操作响应
 */
export interface BatchAccountOperationResponse {
  success: number
  failed: number
  task_ids?: string[]  // 如果是解锁操作，返回任务ID
  errors?: Array<{
    account_id: number
    apple_id: string
    reason: string
  }>
}
